### ISSUE TITLE
Feature Proposal: W3C Context propagation integration via OpenTelemetry

### ISSUE THREAD

**Issue by ikvmw** [rabbitmq/amqp091-go#43]
I notice increasing demand/expectations from user for built-in integration with Distributed tracing.
One such example is Knative where Distributed tracing was integrated from the start. 

Some time ago two major parties - OpenCensus and OpenTracing merged to OpenTelemetry and now Tracing API considered mature - https://github.com/open-telemetry/opentelemetry-go. 

As a first step towards full Distributed tracing support I propose to integrate Tracing context propagation.
 - I suggest to use headers as it is done for example here - https://github.com/knative-sandbox/eventing-rabbitmq/blob/main/cmd/ingress/main.go#L153 to carry context information.
 - I suggest to have same header names across all officially supproted RabbitMQ clients.
 - I argue that merely providing the header naming suggestion in the official docs will be a great step forward as it sets the ground for future development and solves compatibility issues immediately.

Feedback is much appreciated!

**Comment by ikvmw** [comment]
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md

**Comment by lhoguin** [comment]
I think it's fair to start with standardising a header name or two for this purpose. But I would like to know how that would look like across all the protocols we currently support (AMQP0.9.1, AMQP1.0, MQTT, STOMP and Stream).

**Comment by Zerpet** [comment]
Related to #22

**Comment by BryanEllington** [comment]
Also see https://www.w3.org/TR/trace-context/ and https://w3c.github.io/trace-context-amqp/

**Comment by ikvmw** [comment]
with this addition https://github.com/rabbitmq/amqp091-go/pull/96 publishing interface becomes really nice - we don't have to use carriers/propagator concept for publishing  and just pass our context.

**Comment by ikvmw** [comment]
I don't think we need something special here. IIRC by default the propagator is no-op.

**Comment by Zerpet** [comment]
I'm not following. If the propagator is a no-op, how are we going to achieve context propagation, as per [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#context-propagation)?

**Comment by ikvmw** [comment]
by configuring OT SDK to use "proper" propagator :-) For example: https://github.com/open-telemetry/opentelemetry-go/blob/e6839571d26cfc6458050da275f79618fbfd4d9d/propagation.go#L30. It's up to our users to set configure. We going to use only API part.

**Comment by Zerpet** [comment]
Sure, I'm not suggesting we use the SDK and configure a specific implementation. My question is whether we are simply going to mimic [KNative behaviour](https://github.com/knative-sandbox/eventing-rabbitmq/blob/main/cmd/ingress/main.go#L153) (link from OP). In other words, what KNative does is:

- Obtain a span from the context ([link](https://github.com/knative-sandbox/eventing-rabbitmq/blob/b5262252366582b3e5978dbb31e9038591896cf0/cmd/ingress/main.go#L144-L145)) (it is a no-op span if context doesn't have a Span)
- Inject the Span attributes as message headers (see [this](https://github.com/knative-sandbox/eventing-rabbitmq/blob/b5262252366582b3e5978dbb31e9038591896cf0/cmd/ingress/main.go#L153) and [this](https://github.com/knative-sandbox/eventing-rabbitmq/blob/b5262252366582b3e5978dbb31e9038591896cf0/cmd/ingress/main.go#L165-L175))

As a first step, is our intention to do the same in this client?

Are we going to introduce the headers suggested in [this spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#context-propagation)?

**Comment by maxim-badarau-m10** [comment]
Hello, any updates here?

**Comment by lukebakken** [comment]
@maxim-badarau-m10 this issue is still open, so no, it has not been implemented. There is _no reason_ to ask if there are updates.

If this is a critical feature, please contribute to this library or indicate that you have a RabbitMQ support agreement.
